[SPARK-13436][SPARKR] Added parameter drop to subsetting operator [

olarayej opened the pull request:
Added parameter drop to subsetting operator [. This is useful to get a Column from a DataFrame, given its name. R supports it. 

In R:

```
> name <- "Sepal_Length"
> class(iris[, name])
[1] "numeric"
```

Currently, in SparkR:

```
> name <- "Sepal_Length"
> class(irisDF[, name])
[1] "DataFrame"
```

Previous code returns a DataFrame, which is inconsistent with R's behavior. SparkR should return a Column instead. Currently, in order for the user to return a Column given a column name as a character variable would be through `eval(parse(x))`, where x is the string `"irisDF$Sepal_Length"`. That itself is pretty hacky. `SparkR:::getColumn()`is another choice, but I don't see why this method should be externalized. Instead, following R's way to do things, the proposed implementation allows this:

```
> name <- "Sepal_Length"
> class(irisDF[, name, drop=T])
[1] "Column"

> class(irisDF[, name, drop=F])
[1] "DataFrame"
```

This is consistent with R:

```
> name <- "Sepal_Length"
> class(iris[, name])
[1] "numeric"
> class(iris[, name, drop=F])
[1] "data.frame"
```

Commit:
Added parameter validations for colnames<-

Commit:
Added one test for replacing . with _ in column names assignment

Commit:
Added drop parameter to subsetting operator. Rewrote [ as one single method

Commit:
Merge branch 'master' of https://github.com/apache/spark into SPARK-13436

SparkQA commented:
**[Test build #51719 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51719/consoleFull)** for PR 11318 at commit [`632a81f`](https://github.com/apache/spark/commit/632a81fe89b0f461b8084f6f2048046eb17c04b0).

SparkQA commented:
**[Test build #51719 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51719/consoleFull)** for PR 11318 at commit [`632a81f`](https://github.com/apache/spark/commit/632a81fe89b0f461b8084f6f2048046eb17c04b0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51719/
Test PASSed.

gatorsmile commented:
Please update your PR title. Thanks!

olarayej commented:
@felixcheung @sun-rui @shivaram Can you folks please take a look at this one? Thank you!

shivaram commented on R/pkg/R/DataFrame.R:
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
 
 #' @rdname subset
```
can you add some documentation on what `drop` does in this context ?

olarayej replied:
Done!

shivaram commented on R/pkg/R/DataFrame.R:
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
-              j <- as.list(j)
+setMethod("[", signature(x = "DataFrame"),
+          function(x, i, j, ..., drop=T) {
+            # Perform filtering first if needed
```
Also I think `drop=F` should be default if we want to maintain backwards compatibility.

sun-rui replied:
Maybe for Spark 2.0 we can sacrifice some backwards compatibility and have the R user convenience as long term benefit?

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
+            } else {
+              if (class(i) != "Column") {
+                stop(paste0("Expressions other than filtering predicates are not supported ",
+                      "in the first parameter of subsetting operator [."))
```
suggest `extract operator [ or subset()`, as [ is referred to as 'extract' [here](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Extract.html), and subset() calls also go here

felixcheung commented:
Hmm, just as #11336 wouldn't it be confusing then to return a Column, particularly since Column in SparkR is pretty restricted, unlike in R?

sun-rui commented:
https://stat.ethz.ch/R-manual/R-devel/library/base/html/Extract.data.frame.html here for reference

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1271,12 +1280,8 @@ setMethod("[", signature(x = "DataFrame", i = "Column"),
-            } else {
-              x[subset, select, ...]
-            }
+          function(x, subset, select, drop=T, ...) {
```
inconsistent with drop = FALSE in the signature of subset() in the R base package?

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1217,29 +1217,38 @@ setMethod("[[", signature(x = "DataFrame", i = "numericOrcharacter"),
-            if (length(j) > 1) {
-              j <- as.list(j)
+setMethod("[", signature(x = "DataFrame"),
+          function(x, i, j, ..., drop=T) {
```
coding style:
  drop = T

Many other places have such coding style issue too

Commit:
Added docs on drop and fixed style issues.

SparkQA commented:
**[Test build #53357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53357/consoleFull)** for PR 11318 at commit [`6ebe948`](https://github.com/apache/spark/commit/6ebe94846d3881c976102842f1b595a54e2a78fb).

Commit:
pkg/R/DataFrame.R

SparkQA commented:
**[Test build #53359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53359/consoleFull)** for PR 11318 at commit [`adc977f`](https://github.com/apache/spark/commit/adc977f3f3a2acc5de02b3adb6fdd1705843d8dd).

Commit:
Merged unit tests

SparkQA commented:
**[Test build #53360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53360/consoleFull)** for PR 11318 at commit [`017fcdb`](https://github.com/apache/spark/commit/017fcdb8f53efcd011c4956afcd1d543ea344751).

olarayej commented:
@felixcheung @shivaram @sun-rui I have addressed all your comments. Do we have a consensus on the default value for drop? I'd say drop=T makes sense cuz R does it that way. Anyways, it's just a one-line change for me. Please let me know! Thanks!

SparkQA commented:
**[Test build #53360 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53360/consoleFull)** for PR 11318 at commit [`017fcdb`](https://github.com/apache/spark/commit/017fcdb8f53efcd011c4956afcd1d543ea344751).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53360/
Test PASSed.

SparkQA commented:
**[Test build #53359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53359/consoleFull)** for PR 11318 at commit [`adc977f`](https://github.com/apache/spark/commit/adc977f3f3a2acc5de02b3adb6fdd1705843d8dd).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53359/
Test FAILed.

SparkQA commented:
**[Test build #53357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53357/consoleFull)** for PR 11318 at commit [`6ebe948`](https://github.com/apache/spark/commit/6ebe94846d3881c976102842f1b595a54e2a78fb).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53357/
Test PASSed.

sun-rui commented:
@olarayej, there are still similar coding style issues:)

SparkQA commented:
**[Test build #53689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53689/consoleFull)** for PR 11318 at commit [`0473d03`](https://github.com/apache/spark/commit/0473d0325e88c7897de98dd6ce38a5c7c0d38f95).

Commit:
Fixed style issues

SparkQA commented:
**[Test build #53689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53689/consoleFull)** for PR 11318 at commit [`0473d03`](https://github.com/apache/spark/commit/0473d0325e88c7897de98dd6ce38a5c7c0d38f95).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53689/
Test FAILed.

Commit:
pkg/R/DataFrame.R

SparkQA commented:
**[Test build #53691 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53691/consoleFull)** for PR 11318 at commit [`b576fba`](https://github.com/apache/spark/commit/b576fbaddf58a7faa16a5f4cd5046e7c7f33caed).

SparkQA commented:
**[Test build #53691 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53691/consoleFull)** for PR 11318 at commit [`b576fba`](https://github.com/apache/spark/commit/b576fbaddf58a7faa16a5f4cd5046e7c7f33caed).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53691/
Test FAILed.

Commit:
Fixed style issues

SparkQA commented:
**[Test build #53693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53693/consoleFull)** for PR 11318 at commit [`13dfcbb`](https://github.com/apache/spark/commit/13dfcbb68da2692152df5134d3af1993488fa8d4).

SparkQA commented:
**[Test build #53693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53693/consoleFull)** for PR 11318 at commit [`13dfcbb`](https://github.com/apache/spark/commit/13dfcbb68da2692152df5134d3af1993488fa8d4).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53693/
Test FAILed.

Commit:
Fixed style issues

SparkQA commented:
**[Test build #53696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53696/consoleFull)** for PR 11318 at commit [`9ab303a`](https://github.com/apache/spark/commit/9ab303aa639c21398a5bcde139e13214e1c5a6cb).

SparkQA commented:
**[Test build #53696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53696/consoleFull)** for PR 11318 at commit [`9ab303a`](https://github.com/apache/spark/commit/9ab303aa639c21398a5bcde139e13214e1c5a6cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53696/
Test PASSed.

olarayej commented:
@sun-rui Done with the style issues. Thanks!

olarayej commented:
@sun-rui @shivaram Shall we merge this?

Commit:
Merged with latest base branch. Added test cases for drop

SparkQA commented:
**[Test build #57060 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57060/consoleFull)** for PR 11318 at commit [`1a4ca8b`](https://github.com/apache/spark/commit/1a4ca8bd183d559f0560e4e5528f0e44f5791fd7).

SparkQA commented:
**[Test build #57060 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57060/consoleFull)** for PR 11318 at commit [`1a4ca8b`](https://github.com/apache/spark/commit/1a4ca8bd183d559f0560e4e5528f0e44f5791fd7).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57060/
Test FAILed.

Commit:
Fixed style issues

SparkQA commented:
**[Test build #57063 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57063/consoleFull)** for PR 11318 at commit [`b169e14`](https://github.com/apache/spark/commit/b169e14b0e8b722ac71a134331004e6cd3577d13).

SparkQA commented:
**[Test build #57063 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57063/consoleFull)** for PR 11318 at commit [`b169e14`](https://github.com/apache/spark/commit/b169e14b0e8b722ac71a134331004e6cd3577d13).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57063/
Test FAILed.

Commit:
Removed blank space

SparkQA commented:
**[Test build #57146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57146/consoleFull)** for PR 11318 at commit [`def5987`](https://github.com/apache/spark/commit/def59877316a3e9ed2ef10ca428fbc5a6d41e484).

SparkQA commented:
**[Test build #57146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57146/consoleFull)** for PR 11318 at commit [`def5987`](https://github.com/apache/spark/commit/def59877316a3e9ed2ef10ca428fbc5a6d41e484).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57146/
Test FAILed.

SparkQA commented:
**[Test build #57150 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57150/consoleFull)** for PR 11318 at commit [`fc07e85`](https://github.com/apache/spark/commit/fc07e85bd72a5d1dd905c607b27f0eeb05ad6712).

Commit:
Fixed histogram to use drop=F

SparkQA commented:
**[Test build #57150 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57150/consoleFull)** for PR 11318 at commit [`fc07e85`](https://github.com/apache/spark/commit/fc07e85bd72a5d1dd905c607b27f0eeb05ad6712).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57150/
Test FAILed.

Commit:
Fixed style

SparkQA commented:
**[Test build #57151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57151/consoleFull)** for PR 11318 at commit [`290a2cf`](https://github.com/apache/spark/commit/290a2cf4fbb77ffea66650a988c32901f1f29f29).

SparkQA commented:
**[Test build #57151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57151/consoleFull)** for PR 11318 at commit [`290a2cf`](https://github.com/apache/spark/commit/290a2cf4fbb77ffea66650a988c32901f1f29f29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57151/
Test PASSed.

olarayej commented:
@shivaram @sun-rui @felixcheung This one's ready. Shall we merge?

shivaram commented:
Thanks @olarayej for updating this. I think the code looks good. The only thing I'm not sure of is the default value for `[` - I see that you have gone through SparkR's internal code base to add `drop = F`. The reason I prefer keep `drop=F` as the default is that I'm not sure if we have missed some place where we use `[`. While I see the argument that its the default in normal R, I feel like it might not be worth changing the default if we'll run into other unexpected bugs.

SparkQA commented:
**[Test build #57187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57187/consoleFull)** for PR 11318 at commit [`36163e1`](https://github.com/apache/spark/commit/36163e1ad79ad2453e07bac910975a169cd44e81).

Commit:
Changed default value drop=F

SparkQA commented:
**[Test build #57187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57187/consoleFull)** for PR 11318 at commit [`36163e1`](https://github.com/apache/spark/commit/36163e1ad79ad2453e07bac910975a169cd44e81).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57187/
Test PASSed.

olarayej commented:
@shivaram I've changed default value to drop=F as you suggested. Thanks!

shivaram commented:
LGTM. Merging this. Thanks @olarayej
